docs/src/proposals/scrambled-tx.md

```diff
@@ -0,0 +1,335 @@
+fee payer to avoid any kind of fingerprinting while allowing garbage-based
+spamming with amplification risk.
+
+Also, scrambled transaction expires very quickly based on wall clock time, less
+than the targeted duration of difficulty to protect from processing stalls by
+malicious leaders for MEV.
+
+So, scrambled transaction binary format will be like this
+
+```
+[scrambled transaction marker]
+[seed to recursive sha256]
+[recursion count]
+aes256-ed..
+  [aes256 key used to scramble] # this works as authentication tag
+  [expiration in unix time]
+  [normal transaction details...] # fee payer sig cover the scrambling key to avoid unauthorized scrambling
+```
+
+So, this assumes quic to bypass the transaction udp packet size limit.
+
+The number of sha256 recursion is chosen so that anyone can't reveal the
+transaction details, until the cluster optimistically commits the ordering of
+transactions in a block with very high certainty. So, this assumes the recursive
+sha256 is exhibits rather uniform throughput across various hardware (cpu/gpu)
+on the market.
+
+For instance, difficulty can be adjusted 6 9s of availability of MEV-immunity
+status based on standard deviation of past conf time. or hard-coded initially.
+
+Scrambling is optional. While execution of not-scrambled transaction is stalled
+still, infallible and/or MEV-free simple transactions can opt not to be
+scrambled because scrambling might not be desirable for all of situations (HD
+wallets and program deploy transactions).
+
+### Scrambling preparation by clients
+
+Clients need to prepare valid scrambling challenge secrets. By definition, they
+have no choice but to compute the recursive sha256, starting from
+locally-available secure random as seed. However, the challenge secrets are
+designed to be pre-computed before the actual transaction construction/signing.
+So, they can prepare those challenge secrets in background and idling manner for
+later use.
+
+### 2 step voting and voting lock-out
+
+pre-vote and vote are expected to be coalesced to avoid doubling the consensus
+communication overhead. This means a single voting transaction will include
+instructions to pre-vote on block N and to vote on block N-1. This shouldn't
+badly affect latency because block time is still 400ms frequency. Also, the
+pre-vote instruction must contain timestamp.
+
+pre-vote and vote on block N must occur in order. And voting rules including
+lockout are generalized as if any vote comprises of the 2 steps. So fork-choice
+and related subsystem isn't changed significantly.
+
+### Latency analysis
+
+Scrambling will inevitably increases latency. However this should be tolerable
+in trade for the MEV immunity.
+
+Descrambling difficulty can be targeted to be 2x of current opt-conf duration as
+a very crude approach. The 2x is needed as a safe buffer (TODO: needs proper
+statistical analysis) for variation of sha256 throughput across hardware.
```

Comment:
Based on what I know about PoW it's pretty much necessary to assume that centralization of hash power will take place either right away or eventually (I think that'll happen for Bitcoin too),

so related thing to think through is: would there be an incentive for a single entity (that presumably controls the majority of hash power) to speed up and slow down his hash rate (eg. due to price of SOL changing quite rapidly between bull/bear markets), and if so how that would affect difficulty adjustment (would it cause significant chain stall ?).

Reply:
centralization of descrambling power won't happen to the extent of stake centralization.

again, the original prose is quite outdated. but i think there's subtle incentive for non-leader validators to work on descrambling collaboratively. That's beucause leader scheduler is done. everyone is encouraged to make progress distributed descrambling just to antipanate the next block could be their leader block.

Reply:
This comment of mine is also based on wrong assumption, corrected [here](https://github.com/solana-labs/solana/pull/23837#discussion_r1347120491).